### PR TITLE
Merge feature/iot_pnp branch into main

### PR DESCRIPTION
**Pull request by jspaith**
Adding [IoT Plug and Play](https://aka.ms/iotpnp) APIs to embedded SDK.

### CODE REVIEW

**Review by danewalton — APPROVED** [review]
🚀 🚢 :shipit:

**Review by hihigupt — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -302,7 +302,7 @@ git clone https://github.com/Microsoft/vcpkg.git
 # (consider this path as PATH_TO_VCPKG)
 cd vcpkg
 # Checkout the vcpkg commit from the vcpkg-commit.txt file (link above)
-git checkout <vcpkg commit>
+git checkout <vcpkg commit>
```
Comment (ahsonkhan) [inline README.md:305]: Revert this change :p

```suggestion
git checkout <vcpkg commit>
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,11 @@
 
 - Added a current depth field to the JSON reader.
 - Added base64 encoding and decoding APIs that accept `az_span`, available from the `azure/core/az_base64.h` header.
+- Public preview version of a new set of APIs to simplify the experience using Azure IoT Plug and Play. To consume this new feature, the `az::iot::hub` CMake target has been updated. The APIs can be found in the header files: `az_iot_hub_client.h` and `az_iot_hub_client_properties.h`.
+- New samples showing how to consume the new IoT Plug and Play APIs:
+  - paho_iot_pnp_component_sample.c
+  - paho_iot_pnp_sample.c
+  - paho_iot_pnp_with_provisioning_sample.c
 
 ### Bug Fixes
```
Comment (ahsonkhan) [inline CHANGELOG.md:19]: Something to consider, once we delete the feature branch (https://github.com/Azure/azure-sdk-for-c/tree/feature/iot_pnp), this changelog entry (which points to it) will start causing link verification failures:
https://github.com/Azure/azure-sdk-for-c/blob/main/CHANGELOG.md#breaking-changes
image


What should we do here? I wonder if we should update the link to point to the release tag instead.

@weshaggard  thoughts?

@jspaith and @danewalton are there any other lingering links to the feature branch in the repo?

File: sdk/docs/iot/README.md [hunk 3]
```diff
@@ -36,7 +36,6 @@ The library targets made available via CMake are the following:
 
 - `az::iot::hub` - For Azure IoT Hub features ([API documentation here](https://azuresdkdocs.blob.core.windows.net/$web/c/az_iot/1.0.0/az__iot__hub__client_8h.html)).
 - `az::iot::provisioning` - For Azure IoT Provisioning features ([API documentation here](https://azuresdkdocs.blob.core.windows.net/$web/c/az_iot/1.0.0/az__iot__provisioning__client_8h.html)).
-- `az::iot::pnp` - **[BETA]** For Azure IoT Plug and Play features ([feature branch here](https://github.com/Azure/azure-sdk-for-c/tree/feature/iot_pnp)).
```
Comment (ahsonkhan) [inline sdk/docs/iot/README.md:39]: Does this close the https://github.com/Azure/azure-sdk-for-c/issues/1825 or are there more things left to clean up there?

File: sdk/inc/azure/az_iot.h [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 #include <azure/iot/az_iot_common.h>
 #include <azure/iot/az_iot_hub_client.h>
+#include <azure/iot/az_iot_hub_client_properties.h>
```
Comment (ahsonkhan) [inline sdk/inc/azure/az_iot.h:20]: 🎉

File: sdk/inc/azure/iot/az_iot_hub_client_properties.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+ * The payload will be of the form:
+ *
+ * @code
+ * "reported": {
+ *     "<component_name>": {
+ *         "__t": "c",
+ *         "temperature": 23
+ *     }
+ * }
+ * @endcode
```
Comment (ahsonkhan) [inline sdk/inc/azure/iot/az_iot_hub_client_properties.h:37]: **Feedback for a separate PR, rather than part of the merge PR:**

I wonder if this renders correctly in doxygen as part of the "brief" which is typically meant to be a sentence.

Maybe this should be a remark or note instead.

File: sdk/inc/azure/iot/az_iot_hub_client_properties.h [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+ */
+AZ_NODISCARD az_result az_iot_hub_client_properties_writer_end_component(
+    az_iot_hub_client const* client,
+    az_json_writer* ref_json_writer);
+
+/**
+ * @brief Begin a property response to a writable property request from the service.
+ *
+ * This API should be used in response to an incoming writable properties. More details can be
+ * found here:
```
Comment (ahsonkhan) [inline sdk/inc/azure/iot/az_iot_hub_client_properties.h:86]: Same here (for a follow-up cleanup PR), does the extra new line after `@brief` let's doxygen know to render it separately? Otherwise, maybe use `@remark` or something else. If it looks fine, then ignore the feedback.

File: sdk/samples/iot/README.md [hunk 7]
```diff
@@ -603,243 +604,62 @@ This section provides an overview of the different samples available to run and
...
+  <details><summary><i>How to interact with the IoT Plug and Play sample:</i></summary>
 
-    <details><summary><b>Device Twin:</b></summary>
-    <p>
-
-    Two device twin properties are supported in this sample:
-    - A desired property named `targetTemperature` with a `double` value for the desired temperature.
-    - A reported property named `maxTempSinceLastReboot` with a `double` value for the highest temperature reached since device boot.
-    <br>
+  The easiest way to interact with this sample from the service side is to use Azure IoT Explorer.  To use the sample:
```
Comment (ahsonkhan) [inline sdk/samples/iot/README.md:614]: I am going to assume these README changes, removals, and `IoT Hub Plug and Play Sample` heading renames are all intentional and correct. I know there were changes in main independent of the feature branch, and merging this README correctly, without loss of content was very tricky. It might be worth doing a careful review :)

File: sdk/inc/azure/iot/az_iot_hub_client_properties.h (hunk 5 above)
Comment (jspaith) [reply sdk/inc/azure/iot/az_iot_hub_client_properties.h:37]: Agree.  #1893 assigned to me tracks.

File: sdk/inc/azure/iot/az_iot_hub_client_properties.h (hunk 6 above)
Comment (jspaith) [reply sdk/inc/azure/iot/az_iot_hub_client_properties.h:86]: Agree.  #1893 assigned to me tracks.

File: sdk/docs/iot/README.md (hunk 3 above)
Comment (jspaith) [reply sdk/docs/iot/README.md:39]: If there's anything else from #1825, I couldn't find it.  I sent a fair amount of time yesterday trying to find broken stuff like this.

File: CHANGELOG.md (hunk 2 above)
Comment (jspaith) [reply CHANGELOG.md:19]: RE lingering links, a quick `gci -Recurse * | Select-String "feature/iot_pnp"` at root of this directory only shows this hit.

#1894 tracks required changelog.md changes, of this is one, but shouldn't block this PR I don't think because I need to modify changelog after text comes anyway.

* I don't have strong opinions on this comment (other than needing to clean it up of course).  We could either delete the whole sentence which is fine I think and not replace it (I don't see other places in changelog.md referencing beta releases) or I could point it at "https://github.com/Azure/azure-sdk-for-c/releases/tag/1.2.0-beta.1" and word-smith that "this is where it came from versus 'will ship`".
* I need to modify this file (post this PR) as we need final release-worthy text that Wellington will give us today.  

All this clearly needs to go in today/tomorrow, pre-release.

File: README.md (hunk 1 above)
Comment (jspaith) [reply README.md:305]: This is like my old nemesis.  I'm going to get rid of my fork and start from scratch because this keeps sneaking back in.  :(.

In any case I'll push this and get it commited *on this PR*, even if it's a gate run for one character because I'm so sick of this and want it to stop.

File: sdk/samples/iot/README.md (hunk 7 above)
Comment (jspaith) [reply sdk/samples/iot/README.md:614]: I'm assuming they're intentional and correct too:).  That said though I'll do another fresh eyes review presently and if anything is broken I'll update *this PR* and re-run gate for it.

Comment (jspaith) [reply sdk/samples/iot/README.md:614]: Re-review for the win.  

The `IoT Hub Plug and Play` were needed as it brings us inline with branding.  But there was an extra white space (ugh) and a duplicated section not related to anything.

The risk of these confusing future merge/git actions means I want this in this PR so we enter main clean.  

Otherwise good and i'll be pushing.

Comment (ahsonkhan) [reply sdk/samples/iot/README.md:615]: Sounds good. As long as we aren't accidentally losing some of the content (that will be hard to trace back), we are good :)

**Review by ahsonkhan — APPROVED** [review]
Looks reasonable
